pkg/graveler/delete_sensor.go

```diff
@@ -0,0 +1,154 @@
+package graveler
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/treeverse/lakefs/pkg/logging"
+)
+
+const (
+	callbackChannelSize = 1000
+	graceDuration       = 30 * time.Second
+)
+
+type DeleteSensorCB func(repositoryID RepositoryID, branchID BranchID, stagingTokenID StagingToken, inGrace bool)
+
+type StagingTokenCounter struct {
+	StagingTokenID StagingToken
+	Counter        int
+}
+
+type stagingTokenData struct {
+	repositoryID   RepositoryID
+	branchID       BranchID
+	stagingTokenID StagingToken
+}
+
+func (s *stagingTokenData) CombinedKey() string {
+	return fmt.Sprintf("%s:%s", s.repositoryID, s.branchID)
+}
+
+type DeleteSensor struct {
+	ctx       context.Context
+	cb        DeleteSensorCB
+	triggerAt int
+	callbacks chan stagingTokenData
+	wg        sync.WaitGroup
+	mutex     *sync.RWMutex
+	// stopped used as flag that the sensor has stopped. stop processing CountDelete.
+	stopped                int32
+	graceDuration          time.Duration
+	branchTombstoneCounter map[string]*StagingTokenCounter
+}
+
+type DeleteSensorOpts func(s *DeleteSensor)
+
+func WithCBBufferSize(bufferSize int) DeleteSensorOpts {
+	return func(s *DeleteSensor) {
+		s.callbacks = make(chan stagingTokenData, bufferSize)
+	}
+}
+
+func WithGraceDuration(d time.Duration) DeleteSensorOpts {
+	return func(s *DeleteSensor) {
+		s.graceDuration = d
+	}
+}
+func NewDeleteSensor(ctx context.Context, triggerAt int, cb DeleteSensorCB, opts ...DeleteSensorOpts) *DeleteSensor {
+	ds := &DeleteSensor{
+		ctx:                    ctx,
+		cb:                     cb,
+		triggerAt:              triggerAt,
+		stopped:                0,
+		graceDuration:          graceDuration,
+		mutex:                  &sync.RWMutex{},
+		branchTombstoneCounter: make(map[string]*StagingTokenCounter),
+		callbacks:              make(chan stagingTokenData, callbackChannelSize),
+	}
+	for _, opt := range opts {
+		opt(ds)
+	}
+	ds.wg.Add(1)
+	go ds.processCallbacks()
+	return ds
+}
+
+func (s *DeleteSensor) isStopped() bool {
+	return atomic.LoadInt32(&s.stopped) == 1
+}
+
+func (s *DeleteSensor) triggerTombstone(ctx context.Context, st stagingTokenData) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	combinedKey := st.CombinedKey()
+	stCounter, ok := s.branchTombstoneCounter[combinedKey]
+	if !ok {
+		stCounter = &StagingTokenCounter{
+			StagingTokenID: st.stagingTokenID,
+			Counter:        1,
+		}
+		s.branchTombstoneCounter[combinedKey] = stCounter
+		return
+	}
+	if stCounter.StagingTokenID != st.stagingTokenID {
+		stCounter.StagingTokenID = st.stagingTokenID
+		stCounter.Counter = 1
+		return
+	}
+	if stCounter.Counter >= s.triggerAt-1 {
```

Comment:
```suggestion
	if stCounter.Counter >= s.triggerAt-1 {
```
This is odd.  If triggerAt is 2, I would not expect the sensor to fire at the first deletion.

Reply:
It won't, but I agree it makes the code unclear, did this to avoid another if.
Changed